ui/app/components/app-form/project-settings.ts

```diff
@@ -0,0 +1,225 @@
+  }
+
+  get authSSH() {
+    return this.authCase === 5;
+  }
+
+  get authBasic() {
+    return this.authCase === 4;
+  }
+
+  get git() {
+    return this.dataSource?.git;
+  }
+
+
+  validateGitUrl() {
+    let gitUrl = new URL(this.project.dataSource.git.url)
+    // If basic auth, match https url
+    if (this.authCase == 4) {
+      if (gitUrl.protocol != 'https:') {
+        this.flashMessages.error('Git url needs to use "https:" protocol');
+        return false;
+      }
+    }
+    // If ssh force users to use a git: url
+    if (this.authCase == 5) {
+      if (gitUrl.protocol != 'git:') {
+        this.flashMessages.error('Git url needs to use "git:" protocol');
+        return false;
+      }
+    }
+    return true;
+  }
+
+  @action
+  setAuthCase(val:any) {
+    this.authCase = val;
+  }
+
+  @action
+  setBasicAuth(path: string, e: any) {
+    if (!this.project.dataSource?.git?.basic) {
+      this.project.dataSource.git.basic = {
+        username: '',
+        password: ''
+      };
+    }
+    this.project.dataSource.git.basic[path] = e.target.value;
+  }
+
+  @action
+  setSshAuth(path: string, e: any) {
+    if (!this.project.dataSource?.git?.ssh) {
+      this.project.dataSource.git.ssh = {
+        user: '',
+        password: '',
+        privateKeyPem: '',
+      };
+    }
+    this.project.dataSource.git.ssh[path] = e.target.value;
+  }
+
+  @action
+  async saveSettings() {
+    if (!this.validateGitUrl()) {
+      return;
+    }
+    let project = this.project;
+    project.dataSource = this.dataSource;
+    let ref = new Project();
+    ref.setName(project.name);
+    let dataSource = new Job.DataSource();
+    let dataSourcePoll = new Project.Poll();
+    dataSourcePoll.setEnabled(project.dataSourcePoll.enabled);
+    dataSourcePoll.setInterval(project.dataSourcePoll.interval);
+    // Git settings
+    let git = new Job.Git();
+    git.setUrl(project.dataSource.git.url);
+    git.setPath(project.dataSource.git.path);
+    git.setRef(project.dataSource.git.ref);
+
+    // Git Authentication settings
+    if (this.authBasic) {
+      let gitBasic = new Job.Git.Basic();
+      gitBasic.setUsername(this.git.basic.username);
+      gitBasic.setPassword(this.git.basic.password);
+      git.setBasic(gitBasic);
+      git.clearSsh();
+    }
+
+    if (this.authSSH) {
+      let gitSSH = new Job.Git.SSH();
+      let encoder = new window.TextEncoder();
+      gitSSH.setPrivateKeyPem(encoder.encode(this.git.ssh.privateKeyPem));
+      gitSSH.setUser(this.git.ssh.user);
+      gitSSH.setPassword(this.git.ssh.password);
+      git.setSsh(gitSSH);
+      git.clearBasic();
+    }
+
+    dataSource.setGit(git);
+    ref.setDataSource(dataSource);
+    ref.setDataSourcePoll(dataSourcePoll);
+    let applist = project.applicationsList.map((app: Application.AsObject) => {
+      return new Application(app);
+    });
+    ref.setApplicationsList(applist);
+    let req = new UpsertProjectRequest();
+    req.setProject(ref);
```

Comment:
/me takes a breath - wow that's quite the workout, eh?

I'm coming at this with zero experience, so if this is a bad idea please feel free to discard it 😂 - but! - have you considered initializing some of these objects in the constructor and then having an action that does the `setFoo` calls on the PB fields?  I'm not sure how that would work in the pre-filled scenario though when you're setting data from the server... I was just wondering if it's possible to reuse the proto structures instead of having the ProjectModel be a separate data structure that fills in the protos on save.

I'm not sure what's possible, and certainly not something that needs fixed in this PR, just something that struck me reading through this.